File: Node/problems/baekjoon_10814/solution.js
const path = require("path")
const fs = require("fs")

const input = fs.readFileSync(
  path.join(__dirname, "input.txt"),
  "utf8",
)
const lines = input.split("\n")

const caseCount = parseInt(lines[0])
const persons = lines
  .slice(1, caseCount + 1)
  .map((line) => {
    const [age, name] = line.split(" ")

    return {
      name,
      age: parseInt(age),
    }
  })

function solution(list) {
  if (list.length <= 1) return list

  const pivot = list[0]
  const left = []
  const right = []

  for (let i = 1; i < list.length; i++) {
    if (list[i].age < pivot.age) {
      left.push(list[i])
    }
    else if (list[i].age === pivot.age) {
      right.push(list[i])
    }
    else {
      right.push(list[i])
    }
  }

  return [...solution(left), pivot, ...solution(right)]
}

const sorted = solution(persons)
for (let i = 0; i < sorted.length; i++) {
  console.log(`${sorted[i].age} ${sorted[i].name}`)
}

File: Python/problems/baekjoon_10814/solution.py
# BAEKJOON 2839
#
# 링크 : https://www.acmicpc.net/problem/10814
#
# - 퀵 정렬 방법은 시간 초과!!

cases = int(input())
items = []

for i in range(cases):
    items.append(input())


def extract(item):
    [age, name] = item.split(" ")

    return {"name": name, "age": int(age)}


def split(data):
    medium = int(len(data) / 2)

    left = data[:medium]
    right = data[medium:]

    return [left, right]


def merge(left, right):
    result = []
    left_index = 0
    right_index = 0

    while left_index < len(left) and right_index < len(right):
        left_value = extract(left[left_index])
        right_value = extract(right[right_index])

        if left_value["age"] < right_value["age"]:
            result.append(left[left_index])
            left_index += 1
        elif left_value["age"] == right_value["age"]:
            result.append(left[left_index])
            left_index += 1
        else:
            result.append(right[right_index])
            right_index += 1

    while left_index < len(left):
        result.append(left[left_index])
        left_index += 1

    while right_index < len(right):
        result.append(right[right_index])
        right_index += 1

    return result


def solution(data):
    if len(data) <= 1:
        return data

    [left, right] = split(data)

    return merge(solution(left), solution(right))


sorted_data = solution(items)
print("\n".join(sorted_data))
